File: src/de/hsh/kevin/controller/TIController.java
package de.hsh.kevin.controller;

import common.engine.components.game.GameEntryPoint;
import common.updates.UpdateCodes;
import de.hsh.kevin.TIFxmlChanger;
import de.hsh.kevin.logic.Score;
import javafx.application.Platform;

import java.util.Observable;
import java.util.Observer;

/**
 * Erstellt den Controller für die Menüs und das Spiel
 * 
 * @author Kevin
 *
 */
public class TIController extends GameEntryPoint {

    private TIFxmlChanger changer;
    private TIGameController game;
    private TIGameOverController gameOver;
    private Score score;

    /**
     * Erstellt den Controller für Tunnel Invader
     * 
     * @param o
     */
    public TIController(Observer o) {
        super(o, UpdateCodes.TunnelInvader.gameName);
        changer = new TIFxmlChanger(this, TIMenuController.fxml, new TIMenuController());
    }

    /**
     * Ändert die Scenen je nach UpdateCode der übergeben wurde
     */
    @Override
    public void update(Observable o, Object arg) {
        if (arg instanceof String) {
            String msg = (String) arg;

            switch (msg) {
            case UpdateCodes.TunnelInvader.playGame:
                score = new Score();
                if (game != null) {
                    game.deleteObservers();
                }
                game = new TIGameController(score);
                changer.changeGameFxml(o, game);
                break;
            case UpdateCodes.TunnelInvader.gameOver:
                if (game != null) {
                    game.deleteObservers();
                    game = null;
                }
                if (gameOver != null) {
                    gameOver.deleteObservers();
                }
                gameOver = new TIGameOverController(score);
                changer.changeGameOverFxml(o, gameOver);
                gameOver.setScore();
                break;
            default:
                changer.changeFxml(o, (String) arg);
                break;
            }
        }
    }

    /**
     * Render: ruft die Render-Methode des SpielControllers auf
     */
    public void render(int fps) {
        if (game != null) {
            Platform.runLater(() -> game.render(fps));
        }
    }

}

File: Games/Kevin/src/de/hsh/kevin/controller/TIController.kt
package de.hsh.kevin.controller

import common.engine.components.game.GameEntryPoint
import common.updates.UpdateCodes
import de.hsh.kevin.TIFxmlChanger
import de.hsh.kevin.logic.Score
import javafx.application.Platform
import java.util.*

/**
 * Erstellt den Controller für die Menüs und das Spiel
 *
 * @author Kevin
 */
class TIController(o: Observer?) : GameEntryPoint(o, UpdateCodes.TunnelInvader.gameName) {
    private val changer: TIFxmlChanger = TIFxmlChanger(this, TIMenuController.fxml, TIMenuController())
    private var game: TIGameController? = null
    private var gameOver: TIGameOverController? = null
    private var score: Score? = null

    /**
     * Ändert die Scenen je nach UpdateCode der übergeben wurde
     */
    override fun update(o: Observable, arg: Any) {
        if (arg is String) {
            when (arg) {
                UpdateCodes.TunnelInvader.playGame -> {
                    score = Score()
                    if (game != null) {
                        game!!.deleteObservers()
                    }
                    game = TIGameController(score)
                    changer.changeGameFxml(o, game)
                }
                UpdateCodes.TunnelInvader.gameOver -> {
                    if (game != null) {
                        game!!.deleteObservers()
                        game = null
                    }
                    if (gameOver != null) {
                        gameOver!!.deleteObservers()
                    }
                    gameOver = TIGameOverController(score)
                    changer.changeGameOverFxml(o, gameOver)
                    gameOver!!.setScore()
                }
                else                               -> changer.changeFxml(o, arg)
            }
        }
    }

    /**
     * Render: ruft die Render-Methode des SpielControllers auf
     */
    override fun render(fps: Int) {
        Platform.runLater {
            if(game != null) {
                game!!.render(fps)
            }
        }
    }

}
